### ISSUE TITLE
ProperEscapingFunction: flag printf() usages for placeholders being incorrectly

### ISSUE THREAD

**Issue by rebeccahum** [Automattic/VIP-Coding-Standards#674]

**Comment by rebeccahum** [comment]
Sigh, didn't mean to hit enter early.

**Comment by jrfnl** [comment]
Notes for implementation:
1. There are more placeholders than just `%s`. Those will also need to be taken into account.
    ```php
    printf('There are %d monkeys in the %s', $num, $location);
    ```
2. Placeholders can be numbered to change the order of the replacements and/or to re-use replacement values multiple times.
    ```php
    printf('There are %2$d monkeys in the %1$s', $location, $num);
    ```
3. Placeholders can have additional specifications. Prototype: ` %[argnum$][flags][width][.precision]specifier`
4. There is a whole range of `*printf()` functions which should possibly be taken into account: `printf()`, `sprintf()`, `fprintf()`, `vprintf()`, `vsprintf()`, `vfprintf()`.
    Mind: some of these take an array of replacements as the second parameter.

Ref: https://www.php.net/manual/en/function.printf

**Comment by rebeccahum** [comment]
> There is a whole range of *printf() functions which should possibly be taken into account: printf(), sprintf(), fprintf(), vprintf(), vsprintf(), vfprintf().

I think we only care about output in this case, so `sprintf()`, `fprintf()` and `vfprintf()`, we wouldn't need to worry about.

**Comment by jrfnl** [comment]
> sprintf() because it is not directly outputting:

> I think we only care about output in this case, so sprintf(), fprintf() and vfprintf(), we wouldn't need to worry about.

That is irrelevant for this sniff, or at least, the other checks in the sniff don't take it into account.

The below code will be flagged by the sniff, even though it is not being output.
```php
$text = '<div class="' . esc_html( $class ) . '">';
```

**Comment by rebeccahum** [comment]
Hmmm, do you think we should change the sniff to only account for `echo` at the beginning of the statement?

**Comment by jrfnl** [comment]
> Hmmm, do you think we should change the sniff to only account for `echo` and `<?=` at the beginning of the statement?

No, I don't.

The standard uses the WPCS `EscapeOutput` sniff to verify that everything that's being output is being escaped.

This sniff is a check that when escaping functions are used, the correct one is used for each context.

In other words, as things are, these two sniffs are complimentary and will enhance each other.

If this sniff was limited to output context only, it would no longer be complimentary, but would become a subset of the WPCS sniff and should be merged into it.

**Comment by rebeccahum** [comment]
Fair point. I've edited the `What code should not be reported as a violation?` section.
